src/main/scala/com/mwz/sonar/scala/scoverage/ScoverageReportParser.scala

```diff
@@ -0,0 +1,64 @@
+/*
+ * Sonar Scala Plugin
+ * Copyright (C) 2018 All contributors
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package com.mwz.sonar.scala.scoverage
+
+import scala.xml.{Node, NodeSeq, XML}
+
+/**
+ *  Scoverage XML reports parser.
+ *
+ *  @author BalmungSan
+ */
+trait ScoverageReportParser extends ScoverageReportParserAPI {
+
+  /** Parses the scoverage report from a file and returns the [[ModuleCoverage]] */
+  override def parse(scoverageReportFilename: String): ModuleCoverage = {
+    val scoverageXMLReport = XML.loadFile(scoverageReportFilename)
+
+    val moduleScoverage = extractScoverageFromNode(scoverageXMLReport)
+
+    val classCoverages = for {
+      classNode <- scoverageXMLReport \\ "class"
+      filename = s"src/main/scala/${classNode \@ "filename"}"
+      classScoverage = extractScoverageFromNode(classNode)
+      lines = for {
+        statement <- classNode \\ "statement"
+        if (!(statement \@ "ignored").toBoolean)
+        linenum = (statement \@ "line").toInt
+        count = (statement \@ "invocation-count").toInt
+      } yield (linenum -> count)
+      linesCoverage = lines.groupBy(_._1).mapValues(_.map(_._2).sum)
+      classCoverage = FileCoverage(classScoverage, linesCoverage)
+    } yield (filename -> classCoverage)
+
+    // merge the class coverages by filename
+    val files = classCoverages.groupBy(_._1).mapValues(_.map(_._2).reduce(_ + _))
```

Comment:
You seem to be using unnamed accessors on tuples quite a lot, It would probably improve readability a bit if you were unpacking tuples using pattern matching e.g. `classCoverages.groupBy { case (filename, _) => filename }`. This is a bit more verbose, but it's so much better to read and easier to understand that e.g. you are grouping by a filename as opposed to trying to figure out what `_1` or `_2` actually stands for.

Reply:
Yeah I agree it would be better